src/rust/engine/src/intrinsics.rs

```diff
@@ -374,35 +399,39 @@ fn download_file_to_digest(
     let key = Key::from_value(args.pop().unwrap()).map_err(Failure::from_py_err)?;
     let digest = context.get(DownloadedFile(key)).await?;
     let gil = Python::acquire_gil();
-    Snapshot::store_directory_digest(gil.python(), &digest).map_err(throw)
+    Snapshot::store_directory_digest(gil.python(), digest).map_err(throw)
   }
   .boxed()
 }
 
 fn path_globs_to_digest(
   context: Context,
-  mut args: Vec<Value>,
+  args: Vec<Value>,
 ) -> BoxFuture<'static, NodeResult<Value>> {
   async move {
-    let val = args.pop().unwrap();
-    let path_globs = Snapshot::lift_path_globs(&val)
-      .map_err(|e| throw(format!("Failed to parse PathGlobs: {}", e)))?;
+    let path_globs = Python::with_gil(|py| {
+      let py_path_globs = args[0].clone_ref(py).into_ref(py);
+      Snapshot::lift_path_globs(py_path_globs)
+    })
+    .map_err(|e| throw(format!("Failed to parse PathGlobs: {}", e)))?;
     let digest = context.get(Snapshot::from_path_globs(path_globs)).await?;
     let gil = Python::acquire_gil();
-    Snapshot::store_directory_digest(gil.python(), &digest).map_err(throw)
+    Snapshot::store_directory_digest(gil.python(), digest).map_err(throw)
   }
   .boxed()
 }
 
 fn path_globs_to_paths(
   context: Context,
-  mut args: Vec<Value>,
+  args: Vec<Value>,
 ) -> BoxFuture<'static, NodeResult<Value>> {
   let core = context.core.clone();
   async move {
-    let val = args.pop().unwrap();
-    let path_globs = Snapshot::lift_path_globs(&val)
-      .map_err(|e| throw(format!("Failed to parse PathGlobs: {}", e)))?;
+    let path_globs = Python::with_gil(|py| {
+      let py_path_globs = args[0].clone_ref(py).into_ref(py);
```

Comment:
It's a bit odd needing to clone the argument as it comes in, since we already have ownership of it. That was the point of the `pop` before.

Reply:
So the goal is to go from `Arc<Py<PyAny>>` to `&PyAny`. `into_ref(py)` consumes `self`, so we need to go from `Arc<Py<PyAny>>` to `Py<PyAny>>` to `&PyAny`.

Dereferencing `args[0]` only gives back `&Py<PyAny>` because it's an `Arc`. Hence the `clone_ref(py)`.

--

I think if we switch to using `Py<PyAny>` (aka `PyObject`) without the wrapping `Value`, we could avoid the `clone_ref()`.

Reply:
Might be able to use `Py::as_ref` instead of `Py::into_ref` to avoid consuming? (Sorry if this drive-by comment is not helpful!)

Reply:
Not at all drive-by! I think I tried using `as_ref()` and it wasn't compiling, but worth another shot now that everything is green. My attempt was when I was at 400 errors I think 😅 (and then once I got to zero, I had 60 new ones from borrow checker kicking in hehe)

Reply:
Ah ha! This works:

```rust
(*args[0]).as_ref(py)
```

That is, you have to explicitly dereference the `Arc<PyObject>`.

Thanks David!